### DISCUSSION TITLE
Ansible-style support

### DISCUSSION THREAD

**Discussion by mihaigalos** [casey/just#1267] [Ideas]
I love `just` because I can template, configure and code recipes in a natural way using any interpreter or shell I want.
I believe it could even be used as an Ansible alternative, that's why I would like to discuss the idea here.

Here's my painpoint with Ansible:
1. clunky way of working because syntax is for configuration, not programming in the classical sense (functions, loops, conditionals, templates, concepts).
1. yamls written need special predefined modules. If I need to write a shell script, I need to wrap i.e. bash in yaml for Ansible. That means I have a hard time independently testing the core logic.
1. no simple way of performing if-else conditionals.
1. configure-friendly but not code-friendly loops in the classical sense (i.e.: no for, while, repeat keywords).
1. difficult to understand and master tags and conditional statements (i.e.: different include and import behavior).

`just` would offer support for many of the above and leave it up to the user which interpreter (python/bash, etc.) they use.

What are your thoughts on allowing a `ssh()` builtin function to run a recipe on a remote?
This implies:
* opening an `ssh` connection to the remote.
* propagating the `just` binary to over it.
* propagating the `justfile` over it.
* staring execution of the recipe.

One would need to specify a list of hosts the builtin^ applies to.

**Comment by casey** [comment]
I often use `just` in an Ansible-like way, and would definitely like to support that use-case as well as possible. There was a similar issue opened a while ago: #358

Usually I do something like this:

```
foo:
  rsync -avz justfile 10.0.0.1:
  ssh 10.0.0.1 just bar

bar:
```

One argument against adding special SSH/Ansible-like functionality is that it's pretty easy to write a recipe that does what you want, and doing it in a recipe means you can customize it however you want.

**Reply by mihaigalos** [reply]
Beautiful, this is exactly what I meant.
So what I'm saying it support a type of ssh loop. You specify the hosts once and then you run the recipe on each item on the recipe.

**Reply by mihaigalos** [reply]
Also, if the support^ is built in, we don't need to `rsync` every time.

**Reply by casey** [reply]
It's a bit verbose, but this can always be done with a recipe:

```
foo:
  #!/usr/bin/env bash
  set -euxo pipefail
  for host in `read hosts.txt`; do
      rsync -avz justfile $host:
      ssh $host just bar
  done
```

Which also is fully customizable.

**Reply by mihaigalos** [reply]
I like it! `bar` can then contain all the steps required for all servers. Naming here can be improved, but the core looks ok imho.

* What are your thoughts on why this^ is not already an Ansible alternative, leaving aside the 3rd party libs for Ansible?
* Or better asked, what do you think it's still lacking that cannot be done through scripting?

**Reply by runeimp** [reply]
As I understand it Ansible doesn't just copy scripts over to the target and then execute them with whatever interpreter they like. It actually creates shell scripts it executes immediately via a feature of SSH which doesn't store the script on the target system. Also there is Windows. Which Ansible supports via WinRS/WinRM to essentially do the same sort of executions. So Ansible will not leave a target system cluttered with files that were not intended to stay on the system. Or even have to worry about file permissions for such scripts in any way. I'm sure there are many other benefits to using Ansible as well.

It has been a few years now since I've used it so I'm not remembering anything else specific. But for lightweight or personal stuff `just` is a great option. On many systems you will likely need to remember to "clean up after yourself". Especially in corporate environments. But that's not so much more to deal with in general.

**Reply by casey** [reply]
See my comment [here](https://github.com/casey/just/issues/358#issuecomment-1186002525) but i think the best way to support this is with an example recipe in the examples directory.
